common/lib/xmodule/xmodule/modulestore/split_mongo/split.py

```diff
@@ -2922,6 +2920,10 @@ def _copy_subdag(self, user_id, destination_version, block_key, source_blocks, d
                 raw=True,
                 block_defaults=new_block.defaults
             )
+            # _new_block doesn't copy the original_usage data needed for library blocks:
+            if new_block.edit_info.original_usage:
+                destination_block.edit_info.original_usage = new_block.edit_info.original_usage
+                destination_block.edit_info.original_usage_version = new_block.edit_info.original_usage_version
```

Comment:
`_new_block` is generally used for what it sounds like, i.e. creating new blocks, not duplicating blocks. So in general it would not expect any `edit_info` to already exist. As far as I can tell, this use here is the only place in the split code where `_new_block` is used for duplicating a block rather than creating a block.

I could allow `_new_block` to accept an `edit_info` parameter:

```python
    def _new_block(self, existing_args, edit_info=None):
        ...
        # If the new block is really a copy of an existing block, there may be some metadata in edit_info worth preserving:
        edit_info = edit_info or EditInfo()
        edit_info_dict = edit_info.to_storable()
        edit_info_dict.update({  # But the following fields are always reset:
            'edited_on': datetime.datetime.now(UTC),
            'edited_by': user_id,
            'previous_version': None,
            'update_version': new_id
        }
        document = {
            'block_type': category,
            'definition': definition_id,
            'fields': block_fields,
            'edit_info': edit_info_dict,
        }
```

@cpennington Do you prefer that? (n.b. This would also have a side effect of preserving the `source_version` as well, which is currently discarded - not sure if that's desired or not)


@antoviaque Ah, good catch with the duplicating blocks issue. That is in fact tricky because the CMS doesn't tell split that it's duplicating a block; the CMS just uses the normal `create_item()` method instead of a `copy()` method. But there's a bigger problem with that case, which is that the duplicated blocks will have unexpected block IDs, since Randomized Content Blocks (RCBs) normally have very specific block IDs for their children but the CMS copy method assigns arbitrary new IDs to the duplicated children. As a result, if you duplicate an RCB, update the associated library, then synchronize the block, rather than updating its children it will delete all of them and re-create them. This could potentially result in a loss of student state/grade data, if it occurred during a live course.

I'm not sure of the best way to fix that. My first thought is to set some sort of flag on RCBs so that when the CMS duplicates an RCB it will not try to duplicate any children (`dont_duplicate_fields = ('children', 'source_library_version')`). The only downside to that approach is that RCBs where content authors have made override edits to the fields of children will not be able to be copied, so the author would have to set the overrides all over again on the duplicated block, after synchronizing it. However, I don't know to what extent overrides are going to be used anyways. We may want to open a separate story for this.